### PR TITLE
Generate NN archive from training configs

### PR DESCRIPTION
**Pull request by jkbmrz**
This PR introduces a new generator class (`luxonis_train.core.Archiver`) enabling automatic generation of NN archives from training configs. This functionality is exposed both through a command-line interface (CLI) and callbacks (with uploading the NN archive to MLFlow). Furthermore, basic test coverage is included to ensure correctness of generation.

The added features are functional but there are still a few open issues so any insights or suggestions regarding the current implementation would be greatly appreciated!

Open issues:
- (**model formats**): Currently, the generator is only able to archive for ONNX model format. Do we want to support it for other formats also (e.g. BLOB, DLC, XML/BIN… )? If so, how to approach models with multiple executables (e.g. the [path parameter](https://github.com/luxonis/luxonis-ml/blob/dev/luxonis_ml/nn_archive/config_building_blocks/base_models/metadata.py) in NN archive `Metadata` class requires a single path to the model executable so I'm not sure what to do in case of multiple)?
- (**is_softmax** parameter, e.g. in `HeadClassification` [class](https://github.com/luxonis/luxonis-ml/blob/10f9a4323c3ac52e946ef4ba5f49f1701ea0b2a7/luxonis_ml/nn_archive/config_building_blocks/base_models/head.py#L80)): Currently, decision if output is already softmaxed is based on iterating ONNX nodes and checking if there is a node named softmax. However, I'm not sure how bulletproof is that and if it is implementable for other model formats. Is there some other way to determine this parameter (OR should maybe even be hard-coded for specific architectures)?
- (**head_outputs** [classes](https://github.com/luxonis/luxonis-ml/blob/dev/luxonis_ml/nn_archive/config_building_blocks/base_models/head_outputs.py)): I’m wondering how to determine which outputs belong to specific head output parameters (cc @conorsim). For example:
  - `yolo_outputs` parameter in `EfficientBBoxHead`. Is it sensible to expect that all outputs of the model always belong here? If not, how to determine the outputs to be listed?
  - `predictions` parameter in `ClassificationHead`. Is it sensible to expect that models with `ClassificationHead` will always only have one output? If not, how to determine which of the outputs is the right one?
  - `boxes` and `scores` parameters in `ObjectDetectionSSD`. How to determine which output belongs to which parameter for an arbitrary SSD network?
- (**tests**): I’ve implemented setup class method that makes a dummy LDF, config, and ONNX model, and runs the archivation. Next, the tests are run to inspect basic characteristics of the constructed archive file. In the end, teardown class method deletes all the constructed files. Is this approach OK or is there some better way to tackle this?

### CODE REVIEW

**Comment by jkbmrz** [comment]
Any ideas why are we getting all these import errors during the checks? We could manually add them all to the `requirements.txt` but its strange because imports fail even for some packages not introduced by my PR (e.g. cv2)

File: luxonis_train/core/archiver.py [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+            head_node = self.lightning_module._modules["nodes"][head_alias]
+            parameters["iou_threshold"] = head_node.iou_thres
+            parameters["conf_threshold"] = head_node.conf_thres
+            parameters["max_det"] = head_node.max_det
+            # head_outputs["n_keypoints"] # TODO: implement
+            # head_outputs["n_prototypes"] # TODO: implement
+            # head_outputs["prototype_output_name"] # TODO: implement
+        elif head_name == "ObjectDetectionSSD":
+            raise NotImplementedError
+            # head_outputs["anchors"] # TODO: implement
```
Comment (conorsim) [inline luxonis_train/core/archiver.py:274]: I don't think we support any architectures that will require anchors at the moment, just FYI

File: luxonis_train/core/archiver.py [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+        elif head_name == "SegmentationHead":
+            parameters["is_softmax"] = self._is_softmax(executable_path)  # TODO: test
+        elif head_name == "BiSeNetHead":
+            parameters["is_softmax"] = self._is_softmax(executable_path)  # TODO: test
+        elif head_name == "ImplicitKeypointBBoxHead":
+            raise NotImplementedError
```
Comment (conorsim) [inline luxonis_train/core/archiver.py:280]: I think we should be able to implement this one now with latest updates. The ONNX for `ImplicitKeypointBBoxHead` should just have one output, which will be `predictions` in NN archive.

File: luxonis_train/nodes/efficient_bbox_head.py [hunk 3]
```diff
@@ -30,6 +30,7 @@ def __init__(
         n_heads: Literal[2, 3, 4] = 3,
         conf_thres: float = 0.25,
         iou_thres: float = 0.45,
+        max_det: int = 300,
```
Comment (conorsim) [inline luxonis_train/nodes/efficient_bbox_head.py:33]: So I think honestly this is a unique parameter of NN archive which is more for depthai or user preference. I don't think it needs to be a head parameter. We could set it as a constant somewhere else and assume users can change it if desired.

CC @kozlov721 for your opinion as well.

**Review by conorsim — COMMENTED** [review]
Nice, thank you! Just adding some comments for now since there's some questions to sort out. Will reply to those separately.

**Comment by conorsim** [comment]
> (model formats): Currently, the generator is only able to archive for ONNX model format. Do we want to support it for other formats also (e.g. BLOB, DLC, XML/BIN… )? If so, how to approach models with multiple executables (e.g. the [path parameter](https://github.com/luxonis/luxonis-ml/blob/dev/luxonis_ml/nn_archive/config_building_blocks/base_models/metadata.py) in NN archive Metadata class requires a single path to the model executable so I'm not sure what to do in case of multiple)?

In my opinion we only need to support ONNX, especially for now. We might still need to address the secondary executable path issue for ONNX as well, but since we don't have the instance segmentation model implemented at the moment, I wouldn't worry about it in this PR.

**Comment by conorsim** [comment]
> (head_outputs [classes](https://github.com/luxonis/luxonis-ml/blob/dev/luxonis_ml/nn_archive/config_building_blocks/base_models/head_outputs.py)): I’m wondering how to determine which outputs belong to specific head output parameters (cc @conorsim). For example:
yolo_outputs parameter in EfficientBBoxHead. Is it sensible to expect that all outputs of the model always belong here? If not, how to determine the outputs to be listed?
predictions parameter in ClassificationHead. Is it sensible to expect that models with ClassificationHead will always only have one output? If not, how to determine which of the outputs is the right one?
boxes and scores parameters in ObjectDetectionSSD. How to determine which output belongs to which parameter for an arbitrary SSD network?

yolo_outputs: Yes, I think is it sensible to expect that all outputs of the model always belong here.

predictions: Yeah this is the "default" configuration in NN archive where there is only one ONNX output. I think it's safe to assume only one ONNX output in this case.

boxes and scores: Let's not worry about this for now since it's not implemented in luxonis-train.

**Comment by conorsim** [comment]
> (tests): I’ve implemented setup class method that makes a dummy LDF, config, and ONNX model, and run the archivation within the. Next, the tests are run to inspect basic characteristics of the constructed archive file. In the end, teardown class method deletes all the constructed files. Is this approach OK or is there some better way to tackle this?

Yeah this seems good to me. It will be nice to also add tests for each type of head in luxonis-train once other questions are resolved.

File: luxonis_train/core/archiver.py (hunk 1 above)
Comment (kozlov721) [reply luxonis_train/core/archiver.py:274]: `ImplicitKeypointBBoxHead` head does require anchors.

File: luxonis_train/core/archiver.py [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+    def _is_softmax_onnx(self, executable_path) -> bool:
+        """Check if ONNX model output is softmaxed.
+
+        @type executable_path: str
+        @param executable_path: Path to model executable file.
+        """
+
+        model = onnx.load(executable_path)
+        for node in model.graph.node:
+            if node.op_type.lower() == "softmax":
```
[the remark this answers is not in the document]
Comment (kozlov721) [reply luxonis_train/core/archiver.py:309]: We could run the ONNX model and check that the outputs sum to 1.  It's also not an optimal solution though.

File: luxonis_train/nodes/efficient_bbox_head.py (hunk 3 above)
Comment (kozlov721) [reply luxonis_train/nodes/efficient_bbox_head.py:33]: We can then say the same about `conf_thres` and `iou_thres` as those are all parameters of `nms`. The reason the `nms` parameters are head parameters is to ensure the same parameters will be used for all metrics / visualizers etc. and we don't need to specify it for each of them separately.

**Comment by kozlov721** [comment]
> > Any ideas why are we getting all these import errors during the checks? Adding the missing packages to the `requirements.txt` might help but its strange because imports fail even for some package imports that existed before my PR (e.g. cv2).
> 
> I think this is because you `import cv2` in the tests you added. And I think OpenCV is not actually a requirement of this repo. But CC @kozlov721 as well.

`opencv-python` should be installed as part of `luxonis-ml`, so I'm not sure what's the issue exactly. We can add it also to requirements of `luxonis-train`, but it's strange.

File: luxonis_train/core/archiver.py (hunk 1 above)
Comment (conorsim) [reply luxonis_train/core/archiver.py:274]: Ah thanks for pointing that out. Actually need to make a change to NN archive then

Comment (conorsim) [reply luxonis_train/core/archiver.py:274]: (But realizing I need all the object detection parameters in the keypoint head anyways in NN archive)

File: luxonis_train/core/archiver.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (conorsim) [reply luxonis_train/core/archiver.py:309]: Yeah that's not a bad idea either. I think we could hardcode into each classification/segmentation head to keep it simple for now though

File: luxonis_train/nodes/efficient_bbox_head.py (hunk 3 above)
Comment (conorsim) [reply luxonis_train/nodes/efficient_bbox_head.py:33]: Oh my bad I didn't realize they're used in visualizers - makes sense. Disregard my comment

File: luxonis_train/core/archiver.py (hunk 2 above)
Comment (jkbmrz) [reply luxonis_train/core/archiver.py]: adding support in commit [3c0ddc3](https://github.com/luxonis/luxonis-train/pull/17/commits/3c0ddc3088aa834f4b070fdc0808805418dfcc91)

File: tests/unittests/test_core/test_archiver.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+                "archive_save_directory": cls.tmp_path,
+            },
+        }
+        cls.config_path = os.path.join(cls.tmp_path, "tmp_config.yaml")
+        with open(cls.config_path, "w") as yaml_file:
+            yaml_str = yaml.dump(config_dict)
+            yaml_file.write(yaml_str)
+
+        # make model
+        model = torchvision.models.mobilenet_v2(pretrained=False)
```
[the remark this answers is not in the document]
Comment (jkbmrz) [reply tests/unittests/test_core/test_archiver.py:78]: I agree, will look into it!

**Comment by kozlov721** [comment]
> I've added `opencv-python` to `requirements.txt` but now other packages are reported missing. I'm not sure why but maybe `luxonis-ml` isn't installed properly. Might it be possible that I disrupted something by listing its dev version in `requirements.txt` [here](https://github.com/luxonis/luxonis-train/blob/0e3b4de0d60ca8e185e2de1e4e7b76c5a5dfa91c/requirements.txt#L4)?

You're right, seems you accidentally removed the `[all]` specifier so it only installs `luxonis-ml[utils]`. Should be fixed. 
There's also `mlflow` package missing; hotfix for now, I'll update `luxonis-ml` so it also installs all optional filesystem packages when using with the `[all]` specifier.

File: luxonis_train/core/archiver.py [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+
+        head_outputs = {}
+        if head_name == "ClassificationHead":
+            head_outputs["predictions"] = self.outputs[0]["name"]  # TODO: test
+        elif head_name == "EfficientBBoxHead":
+            head_outputs["yolo_outputs"] = [
+                output_dict["name"] for output_dict in self.outputs
+            ]  # TODO: test
+        elif head_name == "ObjectDetectionSSD":
+            raise NotImplementedError  # TODO: boxes, scores
```
[the remark this answers is not in the document]
Comment (conorsim) [reply luxonis_train/core/archiver.py]: I also confirmed with @tersekmatija by the way that we don't have plans of supporting MobileNet SSD in luxonis-train

File: tests/unittests/test_core/test_archiver.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jkbmrz) [reply tests/unittests/test_core/test_archiver.py]: added in commit [5e59c3a](https://github.com/luxonis/luxonis-train/pull/17/commits/5e59c3aebb8a17543446b144697f2e47bc0b67af). For now, all tests are based on ClassificationModel. Later, we can extend tests for all other models.

**Comment by jkbmrz** [comment]
Based on the discussion above, two things remain unresolved:

1. Determining `is_softmax` parameter. As suggested by @conorism we could go for:
> 1/ Hardcoded for each architecture/head in this repo or
> 2/ An option in the training config which we can pull
I’d go for the option 1 for now as we could simply implement it by adding a simple enum like: 
```
class ImplementedHeadsSoxtmaxed(Enum):
  ClassificationHead = True/False
  EfficientBBoxHead = None
  ImplicitKeypointBBoxHead = None
  SegmentationHead = True/False
  BiSeNetHead = True/False
```

2. Expanding tests. As wrote @conorsim:
> It will be nice to also add tests for each type of head in luxonis-train once other questions are resolved.
If I understand correctly, tests should generate a model for each of existing head (ClassificationHead, SegmentationHead, BiSeNetHead, EfficientBBoxHead, ImplicitKeypointBBoxHead) and test if head parameters (also `head_outputs`)  are set correctly?

**Comment by conorsim** [comment]
> If I understand correctly, tests should generate a model for each of the existing heads (ClassificationHead, SegmentationHead, BiSeNetHead, EfficientBBoxHead, ImplicitKeypointBBoxHead) and test if head parameters (also head_outputs) are set correctly?

Yeah this is what I meant.

**Review by conorsim — APPROVED** [review]

**Review by kozlov721 — APPROVED** [review]
LGTM

**Pull request merged**